### DISCUSSION TITLE
How to close the mysql connection before event loop is closed?

### DISCUSSION THREAD

**Discussion by houmie** [fastapi/fastapi#6849] [Questions] [answered]
Hello,

I'm using FastAPI with async unit tests. The unit test is passing successfully, but I get every time an ignored exception.

```
Exception ignored in: <function Connection.__del__ at 0x11023a3a0>
Traceback (most recent call last):
  File "/Users/houmie/.pyenv/versions/v391/lib/python3.9/site-packages/aiomysql/connection.py", line 1072, in __del__
  File "/Users/houmie/.pyenv/versions/v391/lib/python3.9/site-packages/aiomysql/connection.py", line 298, in close
  File "/Users/houmie/.pyenv/versions/3.9.1/lib/python3.9/asyncio/selector_events.py", line 700, in close
  File "/Users/houmie/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 746, in call_soon
  File "/Users/houmie/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 510, in _check_closed
RuntimeError: Event loop is closed
```

**Test:**
```
def setup(self):
    metadata.create_all(engine)
    metadata_radius.create_all(engine)

def teardown(self):
    metadata.drop_all(engine)
    metadata_radius.drop_all(engine)
    self.r.flushall()

@pytest.mark.asyncio
    async def test_unconfirmed_account_returns_4007(self):
        async with AsyncClient(app=app, base_url="http://testserver") as ac:
            ac_id = await crud_account.create_account("h@h.de", "pass1")
            await crud_device.create_device("u1", DeviceType.IPHONE, ac_id)
            response = await ac.post(
                "/login/",
                headers={},
                json={},
            )
            assert response.status_code == 400
            assert response.json()["detail"] == ErrorType.ACCOUNT_NOT_CONFIRMED
```

**conftest.py**
```
@pytest.yield_fixture(scope="session")
def event_loop(request):
    """Create an instance of the default event loop for each test case."""
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()
```

**login.py**
```
@router_login.post("/", status_code=200)
async def login(
    request: Request,
    body: Register,
    response: Response,
    email: Optional[str] = Header(...),
    password: Optional[str] = Header(...),
):
    ac = await crud_account.get_account_by_email(email)
    if auth_service.authenticate(ac, password):
        pass
    else:
        response.status_code = status.HTTP_401_UNAUTHORIZED
        return {"detail": ErrorType.NOT_AUTHORISED}
```

**The exception happens when this method is run, but why?**
```
async def get_account_by_email(email: str) -> account:
    query = account.select().where(account.c.email == email)
    if not db.is_connected:
        await db.connect()
    result = await db.fetch_one(query)
    return result
```

It seems I have to close the connection to database before the loop is closed at the end of the unit test. But how do I do this? 

One thing I have noticed is that these methods are only called when using TestClient:
```
@app.on_event("startup")
async def startup():
    await db.connect()

@app.on_event("shutdown")
async def shutdown():
    await db.disconnect()
```
But because I'm using AsyncClient from httpx, these are not called. What can I do please?

**Comment by frankie567** [comment]
You should use a library like [asgi-lifespan](https://github.com/florimondmanca/asgi-lifespan) to trigger the events with httpx. There is a working example in the README.

**Comment by houmie** [comment]
Thank you. I'm confused about the part where he instantiates the whole app in the test fixture. I want to test the real app, I don't want to create a fake one. I don't understand that.

```
@pytest.fixture
async def app():
    async def startup():
        print("Starting up")

    async def shutdown():
        print("Shutting down")

    app = Starlette(
        routes=[Route("/", home)],
        on_startup=[startup],
        on_shutdown=[shutdown]
    )

    async with LifespanManager(app):
        print("We're in!")
        yield app
```

**Comment by frankie567** [comment]
It's for the sake of the example ; in your case, `app` is just your FastAPI app you wrote in your code. You can just import it in your test file 😄

**Comment by houmie** [comment]
Ha, sorry. I'm new to fast api and asyncio. Please be patient with me.

So I added this to `contest.py`

```
from main import app

@pytest.fixture
async def app1():
    async def startup():
        print("Starting up")

    async def shutdown():
        print("Shutting down")

    async with LifespanManager(app):
        print("We're in!")
        yield app


@pytest.fixture
async def client(app1):
    async with AsyncClient(app=app1, base_url="http://testserver") as client:
        print("Client is ready")
        yield client
```

Then I changed my test to:

```
    async def test_unconfirmed_account_returns_4007(self, client):
        ac_id = await crud_account.create_account("h@h.de", "pass1")
        await crud_device.create_device("u1", DeviceType.IPHONE, ac_id)
        response = await client.post( ...
```

Test passes, but still throws the same exception.
Only the following messages are shown:  We're in! and Client is ready.
Which means startup and shutdown are not called.

**Comment by frankie567** [comment]
Well, it should work like this. I've made a REPL to show you on a simple example: https://repl.it/@frankie567/fastapi-issue-2819

Maybe you have code in your tests that try to query the database *after* the shutdown event is passed (the `setup` and `teardown` methods you show in your first post look suspicious to me).

**Comment by houmie** [comment]
Thank you very much for the example. I misunderstood it at first and thought that I had to override/inject the startup/shutdown in the test itself. Now I understand what you mean.

But one issue remains. This is all fine as long as the calls to the app are made through AsyncClient ```await client.post("/", ... ```.

But there are cases where I would like to test an async function that accesses the database, which isn't exposed directly to any router. So I would like to test that function and see that it behaves correctly. What do I do there? Due to lack of Startup/Shutdown event in this scenario, I won't be able to disconnect the database cleanly and it throws the same error as mentioned earlier.

**Comment by frankie567** [comment, marked as answer]
You're welcome!

What you could do to solve this is to create a yieldable fixture for the database. Something like:

```py
@pytest.fixture
async def test_database():
  await database.connect()
  yield database
  await database.disconnect()
```

**Comment by houmie** [comment]
Brilliant. Thank you. I'll close the issue then. It seems that I have it all working now. 👍🏼

**Comment by tiangolo** [comment]
Thanks for the help here @frankie567 ! 👏  🙇 

Thanks for reporting back and closing the issue @houmie  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
